src/Interpreters/ExpressionAnalyzer.cpp

```diff
@@ -1819,7 +1826,10 @@ ExpressionAnalysisResult::ExpressionAnalysisResult(
 
         chain.finalize();
 
-        finalize(chain, prewhere_step_num, where_step_num, having_step_num, query);
+        /// For creating parameterized view, query parameters are allowed in select
+        /// As select will be stored without substituting query parameters, we don't want to evaluate the expressions/steps
+        if (!query_analyzer.query_options.is_create_parameterized_view)
+            finalize(chain, prewhere_step_num, where_step_num, having_step_num, query);
```

Comment:
Not obvious change and looks dangerous  :( 
(this finalize might be needed for part of the query not related to parameters?) 
what kind of error we might get here?

Reply:
The error would be that we are trying to access a step which does exist in the map `chain.steps.at(where_step_num)->required_output`

Suppose we had the query 
`CREATE VIEW v3 AS SELECT * FROM Catalog WHERE Price=20 AND Quantity=3`

The steps would be 
0 : INPUT () (no column) String Name
1 : INPUT () (no column) UInt64 Price
2 : INPUT () (no column) UInt64 Quantity
3 : COLUMN () Const(UInt8) UInt8 20
4 : FUNCTION (1, 3) (no column) UInt8 equals(Price, 20) [equals]
5 : COLUMN () Const(UInt8) UInt8 3
6 : FUNCTION (2, 5) (no column) UInt8 equals(Quantity, 3) [equals]
7 : FUNCTION (4, 6) (no column) UInt8 and(equals(Price, 20), equals(Quantity, 3)) [and]

And in the finalise function we try to find the step of where which would be ‘and(equals(Price, 20), equals(Quantity, 3))’. Now all of this evaluation can be done if we have literals instead of query parameters. Since we have query parameters, we skip adding the steps (near `appendWhere`) & also evaluating the step (`finalise` i.e this function).

Now that our query is 
`CREATE VIEW v3 AS SELECT * FROM Catalog WHERE Price={price:UInt64} AND Quantity={quantity:UInt64}`

The steps would just be
0 : INPUT () (no column) String Name
1 : INPUT () (no column) UInt64 Price
2 : INPUT () (no column) UInt64 Quantity

Even if the query is `CREATE VIEW v3 AS SELECT * FROM Catalog WHERE Price={price:UInt64} AND Quantity=3`, we cannot evaluate it as the where condition would be `and(equals(Price, price), equals(Quantity, 3))`.

I agree with your point that maybe we need to evaluate in other parts of the query, I have moved this function to where section of the finalise function, since this is what applies to parameterized view.
It would be great if we could skip all this evaluation for CREATE VIEW, but as we discussed previously some parts like `getRootActions` are needed so we would have to work around it and skip only the parts that don’t apply.